From 5235f306e75ab5417fb6b1c417aae8d842a354e8 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Thu, 24 Sep 2026 16:10:04 +0800
Subject: [PATCH] mtk: mt76: mt7915: fix the behavior of radar detection

RDD_DET_MODE is a firmware command intended for testing and does not
pause TX after radar detection, so remove it from the normal flow.
Additionally, introduce mt7915_mcu_rdd_resume_tx to resume TX traffic
after radar detection.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 mt7915/mac.c    |  2 +-
 mt7915/main.c   | 11 +++++++++++
 mt7915/mcu.c    | 32 ++++++++++++++++++++++++++++++--
 mt7915/mt7915.h |  5 +++++
 mt7915/vendor.c | 28 ++++++++++++++++++++++++++++
 5 files changed, 75 insertions(+), 3 deletions(-)

diff --git a/mt7915/mac.c b/mt7915/mac.c
index acba41c57..19f9a8b2c 100644
--- a/mt7915/mac.c
+++ b/mt7915/mac.c
@@ -2202,7 +2202,7 @@ static int mt7915_dfs_start_rdd(struct mt7915_dev *dev, int rdd_idx)
 			return err;
 	}
 
-	return mt76_connac_mcu_rdd_cmd(&dev->mt76, RDD_DET_MODE, rdd_idx, 0, 1);
+	return 0;
 }
 
 static int mt7915_dfs_start_radar_detector(struct mt7915_phy *phy)
diff --git a/mt7915/main.c b/mt7915/main.c
index 066f8d85c..103550866 100644
--- a/mt7915/main.c
+++ b/mt7915/main.c
@@ -137,6 +137,7 @@ static void mt7915_stop(struct ieee80211_hw *hw, bool suspend)
 
 	mutex_lock(&dev->mt76.mutex);
 
+	mt7915_mcu_rdd_resume_tx(phy);
 	mt76_testmode_reset(phy->mt76, true);
 
 	clear_bit(MT76_STATE_RUNNING, &phy->mt76->state);
@@ -781,6 +782,10 @@ mt7915_post_channel_switch(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 	struct cfg80211_chan_def *chandef = &phy->mt76->chandef;
 	int ret;
 
+	ret = mt7915_mcu_rdd_resume_tx(phy);
+	if (ret)
+		goto out;
+
 	ret = cfg80211_chandef_dfs_required(hw->wiphy, chandef, NL80211_IFTYPE_AP);
 	if (ret <= 0)
 		goto out;
@@ -1935,6 +1940,12 @@ mt7915_set_radar_background(struct ieee80211_hw *hw,
 
 	mutex_lock(&dev->mt76.mutex);
 
+	/* ignore disable background radar when cert mode is on */
+	if (phy->dfs_tx_mode && !chandef) {
+		ret = 0;
+		goto out;
+	}
+
 	if (dev->mt76.region == NL80211_DFS_UNSET)
 		goto out;
 
diff --git a/mt7915/mcu.c b/mt7915/mcu.c
index 38e52f513..b22a08d63 100644
--- a/mt7915/mcu.c
+++ b/mt7915/mcu.c
@@ -344,15 +344,43 @@ mt7915_mcu_rx_radar_detected(struct mt7915_dev *dev, struct sk_buff *skb)
 	if (!mphy)
 		return;
 
-	if (r->rdd_idx == MT_RDD_IDX_BACKGROUND)
+	if (r->rdd_idx == MT_RDD_IDX_BACKGROUND) {
 		cfg80211_background_radar_event(mphy->hw->wiphy,
 						&dev->rdd2_chandef,
 						GFP_ATOMIC);
-	else
+	} else {
+		struct mt7915_phy *phy = mphy->priv;
+
+		phy->rdd_tx_paused = true;
 		ieee80211_radar_detected(mphy->hw, NULL);
+	}
 	dev->hw_pattern++;
 }
 
+int mt7915_mcu_rdd_resume_tx(struct mt7915_phy *phy)
+{
+#define MAC_DFS_RESUME_TX	2
+	struct {
+		u8 mac_enable;
+		u8 band_idx;
+		u8 rsv[2];
+	} __packed req = {
+		.mac_enable = MAC_DFS_RESUME_TX,
+		.band_idx = phy->mt76->band_idx,
+	};
+	int ret;
+
+	if (!phy->rdd_tx_paused)
+		return 0;
+
+	ret = mt76_mcu_send_msg(&phy->dev->mt76, MCU_EXT_CMD(MAC_INIT_CTRL),
+				&req, sizeof(req), true);
+	if (!ret)
+		phy->rdd_tx_paused = false;
+
+	return ret;
+}
+
 static void
 mt7915_mcu_rx_log_message(struct mt7915_dev *dev, struct sk_buff *skb)
 {
diff --git a/mt7915/mt7915.h b/mt7915/mt7915.h
index 24e85f8ba..732cc540f 100644
--- a/mt7915/mt7915.h
+++ b/mt7915/mt7915.h
@@ -401,6 +401,8 @@ struct mt7915_phy {
 	struct list_head stats_list;
 	spinlock_t stats_lock;
 
+	bool rdd_tx_paused;
+
 #ifdef CONFIG_NL80211_TESTMODE
 	struct {
 		u32 *reg_backup;
@@ -434,6 +436,8 @@ struct mt7915_phy {
 	} csi;
 
 	struct mt7915_air_monitor_ctrl amnt_ctrl;
+
+	u8 dfs_tx_mode;
 #endif
 	struct mt7915_scs_ctrl scs_ctrl;
 #ifdef MTK_DEBUG
@@ -804,6 +808,7 @@ void mt7915_mcu_rx_event(struct mt7915_dev *dev, struct sk_buff *skb);
 void mt7915_mcu_exit(struct mt7915_dev *dev);
 void mt7915_mcu_wmm_pbc_work(struct work_struct *work);
 int mt7915_mcu_enable_obss_spr(struct mt7915_phy *phy, u8 action, u8 val);
+int mt7915_mcu_rdd_resume_tx(struct mt7915_phy *phy);
 
 static inline u16 mt7915_wtbl_size(struct mt7915_dev *dev)
 {
diff --git a/mt7915/vendor.c b/mt7915/vendor.c
index 2cf3941f9..4a21d8998 100644
--- a/mt7915/vendor.c
+++ b/mt7915/vendor.c
@@ -1588,11 +1588,39 @@ static int mt7915_vendor_ibf_ctrl_dump(struct mt76_dev *mdev, bool *enabled)
 	return 0;
 }
 
+static int mt7915_vendor_dfs_tx_mode_ctrl(struct mt76_dev *mdev, u8 radio_idx,
+					  u8 mode)
+{
+	struct mt7915_dev *dev = container_of(mdev, struct mt7915_dev, mt76);
+	struct mt7915_phy *phy = &dev->phy;
+	int rdd_idx, err;
+
+	if (!phy->mt76->cap.has_5ghz)
+		phy = mt7915_ext_phy(dev);
+	if (!phy || !phy->mt76->cap.has_5ghz) {
+		dev_err(dev->mt76.dev, "Cannot find 5G phy\n");
+		return -EINVAL;
+	}
+
+	rdd_idx = mt7915_get_rdd_idx(phy, false);
+	if (rdd_idx < 0) {
+		dev_err(dev->mt76.dev, "Invalid rdd idx %d\n", rdd_idx);
+		return -EINVAL;
+	}
+
+	err = mt76_connac_mcu_rdd_cmd(&dev->mt76, RDD_DET_MODE, rdd_idx, 0, 1);
+	if (!err)
+		phy->dfs_tx_mode = mode;
+
+	return err;
+}
+
 static const struct mt76_vendor_ops mt7915_vendor_ops = {
 	.commands = mt7915_vendor_commands,
 	.n_commands = ARRAY_SIZE(mt7915_vendor_commands),
 	.ibf_ctrl = mt7915_vendor_ibf_ctrl,
 	.ibf_ctrl_dump = mt7915_vendor_ibf_ctrl_dump,
+	.dfs_tx_mode_ctrl = mt7915_vendor_dfs_tx_mode_ctrl,
 };
 
 void mt7915_vendor_register(struct mt7915_phy *phy)
-- 
2.45.2

